resolver/resolver.cc

```diff
@@ -1317,16 +1374,58 @@ class ResolveConstantsWalk {
     }
 
 public:
-    ResolveConstantsWalk() : nesting_(nullptr) {}
+    ResolveConstantsWalk() : nesting_(nullptr), loadScopeDepth_(0) {
+        firstDefinitionLocs = make_unique<UnorderedMap<core::SymbolRef, core::LocOffsets>>();
+    }
 
-    void preTransformClassDef(core::Context ctx, ast::ExpressionPtr &tree) {
-        nesting_ = make_unique<Nesting>(std::move(nesting_), ast::cast_tree_nonnull<ast::ClassDef>(tree).symbol);
+    void preTransformMethodDef(core::Context ctx, ast::ExpressionPtr &tree) {
+        ENFORCE(loadScopeDepth_ >= 0);
+        loadScopeDepth_++;
+    }
+
+    void postTransformMethodDef(core::Context ctx, ast::ExpressionPtr &tree) {
+        ENFORCE(loadScopeDepth_ > 0);
+        loadScopeDepth_--;
+    }
+
+    void preTransformBlock(core::Context ctx, ast::ExpressionPtr &tree) {
+        ENFORCE(loadScopeDepth_ >= 0);
+        loadScopeDepth_++;
+    }
+
+    void postTransformBlock(core::Context ctx, ast::ExpressionPtr &tree) {
+        ENFORCE(loadScopeDepth_ > 0);
+        loadScopeDepth_--;
     }
 
     void postTransformUnresolvedConstantLit(core::Context ctx, ast::ExpressionPtr &tree) {
         walkUnresolvedConstantLit(ctx, tree);
     }
 
+    void preTransformClassDef(core::Context ctx, ast::ExpressionPtr &tree) {
+        auto &original = ast::cast_tree_nonnull<ast::ClassDef>(tree);
+        auto sym = original.symbol;
+
+        // Populate local first definitions table for out-of-order reference checking.
+        // This happens *iff* the symbol is:
+        //   a). defined only in this file and nowhere else
+        //   b). defined multiple times in this file
+        //   c). this particular classDef is the first occurrence of the symbol in the file
+        if (!ctx.file.data(ctx).isRBI() && loadScopeDepth_ == 0 && sym.isOnlyDefinedInFile(ctx.state, ctx.file)) {
+            auto defLoc = sym.data(ctx)->loc().offsets();
+            auto declLoc = original.declLoc;
+
+            if (defLoc.beginPos() >= declLoc.endPos() && firstDefinitionLocs->find(sym) == firstDefinitionLocs->end()) {
+                // When this condition is met, effectively it means the file has multiple definitions of the symbol
+                // and this is the first one. We need to use the first def for out-of-order checking, since any
+                // reference *before* the first def will be out of order.
+                firstDefinitionLocs->insert(std::make_pair(sym, declLoc));
```

Comment:
Given that you're using `insert` here, you can drop the check for `firstDefinitionLocs->find` above because `insert` will only create an entry if none exists:

https://en.cppreference.com/w/cpp/container/unordered_map/insert
https://github.com/abseil/abseil-cpp/blob/0c3df2f5a7d3918b9e5762cd8f143a6004b76cda/absl/container/flat_hash_map.h#L246-L266